miasm2/analysis/mem.py

```diff
@@ -0,0 +1,1259 @@
+          fields.
+
+    Example:
+        class Example(MemStruct):
+            fields = [
+                # Number field: just struct.pack fields with one value
+                ("num", Num("I")),
+                ("flags", Num("B")),
+                # Ptr fields are Num, but they can also be dereferenced
+                # (self.deref_<field>). Deref can be read and set.
+                ("other", Ptr("I", OtherStruct)),
+                ("i", Ptr("I", Num("I"))),
+                # Ptr to a variable length String
+                ("s", Ptr("I", MemStr)),
+            ]
+
+        mstruct = MyStruct(vm, addr)
+
+        # Field assignment modifies virtual memory
+        mstruct.num = 3
+        assert mstruct.num == 3
+        memval = struct.unpack("I", vm.get_mem(mstruct.get_addr(),
+                                                      4))[0]
+        assert memval == mstruct.num
+
+        # Memset sets the whole structure
+        mstruct.memset()
+        assert mstruct.num == 0
+        mstruct.memset('\x11')
+        assert mstruct.num == 0x11111111
+
+        other = OtherStruct(vm, addr2)
+        mstruct.other = other.get_addr()
+        assert mstruct.other == other.get_addr()
+        assert mstruct.deref_other == other
+        assert mstruct.deref_other.foo == 0x1234
+
+    See the various MemField doc for more information.
+    """
+    __metaclass__ = _MetaMemStruct
+
+    fields = []
+
+    _size = None
+
+    # Classic usage methods
+
+    def __init__(self, vm, addr=None, *args, **kwargs):
+        global ALLOCATOR
+        super(MemStruct, self).__init__(*args, **kwargs)
+        self._vm = vm
+        if addr is None:
+            if ALLOCATOR is None:
+                raise ValueError("Cannot provide None address to MemStruct() if"
+                                 "%s.set_allocator has not been called."
+                                 % __name__)
+            self._addr = ALLOCATOR(vm, self.get_size())
+        else:
+            self._addr = addr
+
+    def get_addr(self, field_name=None):
+        """Return the address of this MemStruct or one of its fields.
+
+        Args:
+            field_name: (str, optional) the name of the field to get the
+                address of
+        """
+        if field_name is not None:
+            if field_name not in self._attrs:
+                raise ValueError("This structure has no %s field" % field_name)
+            offset = self._attrs[field_name]['offset']
+        else:
+            offset = 0
+        return self._addr + offset
+
+    @classmethod
+    def sizeof(cls):
+        """Return the static size of this structure, when available (it is the
+        case by default).
+        """
+        # Child classes can set cls._size if their size is not the sum of
+        # their fields
+        if cls._size is None:
+            return sum(a["field"].size() for a in cls._attrs.itervalues())
```

Comment:
You may set `cls._size`, giving a lazy `sizeof` (it seems that this size is constant)